packages/core/src/lib/filter/subscription.ts

```diff
@@ -0,0 +1,286 @@
+import { Stream } from "@libp2p/interface/connection";
+import {
+  CustomEvent,
+  EventEmitter,
+  EventHandler
+} from "@libp2p/interface/events";
+import type { Peer } from "@libp2p/interface/peer-store";
+import type {
+  ContentTopic,
+  IDecodedMessage,
+  IDecoder,
+  IProtoMessage,
+  PubSubTopic,
+  SubscriptionEventMap
+} from "@waku/interfaces";
+import { WakuMessage } from "@waku/proto";
+import { groupByContentTopic, Logger } from "@waku/utils";
+import all from "it-all";
+import * as lp from "it-length-prefixed";
+import { pipe } from "it-pipe";
+
+import { FilterSubscribeResponse, FilterSubscribeRpc } from "./filter_rpc.js";
+
+const log = new Logger("waku:filter-v2:subscription");
+
+// one subscription object per pubsub topic
+export class Subscription extends EventEmitter<SubscriptionEventMap> {
```

Comment:
I'd like to avoid `extends EventEmitter` and prefer [this way](https://github.com/waku-org/waku-frontend/blob/f7ba689d8796c8693595c13d8df26453bfb719c5/src/services/waku.ts#L23) when `emitter` is proxied by appropriate functions.

Reply:
These are good points and let's keep it using for now, created task for moving away from it https://github.com/waku-org/js-waku/issues/1737

My comment here was more about not `extending` the class and creating proxy functions. This is not that convenient but it brings a bunch of benefits. As `Subscription` is internal class it is not that important but overall I think we should follow with practice.

Reply:
I think for general overview you can just search but for our case most important would be ability to define behavior/side effects and avoid collision with base classes. 

So in case of `Subscription` we want to have `add/remove` event listeners but don't want to have `dispatch` exposed.

Reply:
67b7c1ef8b0c6738fd38bbcab61752d970079f5c